File: 629. K Inverse Pairs Array/kInversePairs.cpp
class Solution
{
public:
    int dp[1001][1001];
    int mod = 1e9 + 7;
    int solve(int n, int k)
    {
        if (n == 0)
            return 0;
        if (k == 0)
            return 1;
        if (dp[n][k] != -1)
            return dp[n][k];

        int res = 0;
        for (int inv = 0; inv <= min(k, n - 1); inv++)
        {
            res = (res + solve(n - 1, k - inv)) % mod;
        }
        return dp[n][k] = res;
    }
    int kInversePairs(int n, int k)
    {
        memset(dp, -1, sizeof(dp));
        return solve(n, k);
    }
};

File: 629. K Inverse Pairs Array/kInversePairs.java
class Solution {
    public int kInversePairs(int n, int k) {

        int dp[][] = new int[n + 1][k + 1];
        int mod = 1000000007;
        // dp[i][j] = total number of arrays size from 1 to i in k inversions

        // if array size is zero then we don't have any inversions
        // if we have array size and inversion 0
        for (int i = 0; i <= n; i++) {
            dp[i][0] = 1;
        }

        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= k; j++) {
                for (int inv = 0; inv <= Math.min(j, i - 1); inv++) {
                    dp[i][j] = (dp[i][j] + dp[i - 1][j - inv]) % mod;
                }
            }
        }

        return dp[n][k];
    }
}